### PR TITLE
Fix smoke tests

### PR DESCRIPTION
**Pull request by yarpo**
### Change description
Changed IDAM user authentication

### Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,3863 @@
+{
```
Comment (timja-kainos) [inline package-lock.json]: remove this file please

File: package.json [hunk 2]
```diff
@@ -77,6 +77,7 @@
     "mime": "^2.0.3",
     "moment": "^2.18.1",
     "nodelist-foreach-polyfill": "^1.2.0",
+    "npm": "^5.8.0",
```
Comment (timja-kainos) [inline package.json]: why this change?

File: src/integration-test/helpers/clients/idamClient.ts [hunk 3]
```diff
@@ -4,6 +4,12 @@ const baseURL: string = process.env.IDAM_URL
 
 const defaultPassword = 'Password12'
 
+const oauth2 = {
+  client_id: 'cmc_citizen',
+  redirect_uri: 'https://localhost:3000/receiver',
```
Comment (timja-kainos) [inline src/integration-test/helpers/clients/idamClient.ts]: the redirect_uri and client_secret need to be brought in as environment variables like @sabahirfan did in the claim-store change

File: src/integration-test/bootstrap/bootstrap.ts [hunk 4]
```diff
@@ -70,16 +71,15 @@ async function waitTillHealthy (appURL: string) {
 
 async function createSmokeTestsUserIfDoesntExist (username: string, userGroup: string, password: string): Promise<void | string> {
   try {
-    return await IdamClient.authenticateUser(username, password)
+    const token = await IdamClient.authenticateUser(username, password)
+
+    return `Bearer ${token}`
```
Comment (geek-soft) [inline src/integration-test/bootstrap/bootstrap.ts]: That change is not needed. What is more it makes more sense to not return from `createSmokeTestsUserIfDoesntExist` as return value is never consumed.

File: src/integration-test/helpers/claimStoreHelper.ts [hunk 5]
```diff
@@ -16,8 +15,9 @@ class ClaimStoreHelper extends codecept_helper {
     const claim: Claim = await ClaimStoreClient.retrieveByReferenceNumber(claimRef, { bearerToken: claimantBearerToken })
     const pinResponse = await IdamClient.getPin(claim.letterHolderId)
     const upliftToken = await IdamClient.authenticatePinUser(pinResponse.body)
-    const defendantBearerToken = await IdamClient.upliftUser(defendantEmail, upliftToken)
-    await ClaimStoreClient.linkDefendant({ bearerToken: defendantBearerToken }, claim.externalId)
+    await IdamClient.upliftUser(defendantEmail, upliftToken)
```
Comment (geek-soft) [inline src/integration-test/helpers/claimStoreHelper.ts]: That does not look right. It looks like you start uplift process without completing it and then in next two lines you just link newly created defendant, not the one that got created in uplift process.

File: src/integration-test/bootstrap/bootstrap.ts (hunk 4 above)
Comment (sabahirfan) [reply src/integration-test/bootstrap/bootstrap.ts]: done

File: src/integration-test/helpers/clients/idamClient.ts [hunk 6]
```diff
@@ -31,6 +37,43 @@ export class IdamClient {
...
+  /**
+   * Authenticate user
+   *
+   * @param {string} email
+   * @param password the users password (optional, default will be used if none provided)
+   * @returns {Promise<string>}
+   */
+  static async authenticateUser (username: string, password: string = undefined): Promise<string> {
+
+    password = password || defaultPassword
```
Comment (geek-soft) [inline src/integration-test/helpers/clients/idamClient.ts]: Please do not resign password argument. I preferred `${username}:${password || defaultPassword}`.

File: src/integration-test/helpers/clients/idamClient.ts [hunk 7]
```diff
@@ -31,6 +37,43 @@ export class IdamClient {
...
+
+    const base64Authorisation: string = IdamClient.toBase64(`${username}:${password}`)
+    const oauth2Params: string = IdamClient.toUrlParams(oauth2)
+
+    const authResponse = await request.post({
+      url: `${baseURL}/oauth2/authorize?response_type=code&${oauth2Params}`,
```
Comment (geek-soft) [inline src/integration-test/helpers/clients/idamClient.ts]: There could be also one private `authorize` method that does most of the heavy lifting  which is called by `authenticateUser`, `authenticatePinUser` and `upliftUser` method as all these methods needs OAuth2 query parameters.

File: src/integration-test/helpers/clients/idamClient.ts (hunk 6 above)
Comment (yarpo) [reply src/integration-test/helpers/clients/idamClient.ts]: done

[but this is typical JS solution for args with default value]

File: src/integration-test/helpers/clients/idamClient.ts [hunk 8]
```diff
@@ -31,6 +37,43 @@ export class IdamClient {
...
+      headers: { Authorization: `Basic ${base64Authorisation}` }
+    })
+
+    const tokenParams = IdamClient.toUrlParams({
```
[the remark this answers is not in the document]
Comment (yarpo) [reply src/integration-test/helpers/clients/idamClient.ts]: done

File: docker-compose.smoke-tests.yml [hunk 9]
```diff
@@ -11,6 +11,7 @@ services:
         - https_proxy
         - no_proxy=remote-webdriver
         - IDAM_URL
+        - OAUTH2_CLIENT_SECRET
```
Comment (geek-soft) [inline docker-compose.smoke-tests.yml]: The same has to be done in `cmc-integration-tests` repo

Comment (yarpo) [reply docker-compose.smoke-tests.yml]: done
https://github.com/hmcts/cmc-integration-tests/pull/176
